agent/src/capability/servicemanifestfileloader.cpp

```diff
@@ -0,0 +1,144 @@
+    ServiceManifestLoader(source)
+{
+}
+
+std::vector<json_t *> ServiceManifestFileLoader::loadContent()
+{
+    std::string errorMessage;
+
+    if (m_source.empty()) {
+        //File path is empty, this is ok
+        log_debug() << "Path to service manifests is empty";
+    } else if (m_source.compare("/") == 0) {
+        errorMessage = "Searching for configuration files from root dir not allowed";
+        log_error() << errorMessage;
+        throw ServiceManifestPathError(errorMessage);
+    } else if (isDirectory(m_source)) {
+        log_debug() << "Path to service manifests is a dir";
+        loadServiceManifestDir();
+    } else if (isFile(m_source)) {
+        log_debug() << "Path to service manifest is a file";
+        loadServiceManifestFile(m_source);
+    } else {
+        errorMessage = "Path to service manifest does not exist: \"" + m_source + "\"";
+        log_error() << errorMessage;
+        throw ServiceManifestPathError(errorMessage);
+    }
+    return m_content; // TODO: is this a good way to handle the default?
+}
+
+void ServiceManifestFileLoader::loadServiceManifestDir()
+{
+    std::string errorMessage;
+
+    if (isDirectoryEmpty(m_source)) {
+        log_warning() << "Path to configuration files is empty: " << m_source;
+        return;
+    }
+
+    std::vector<std::string> files = fileList();
+
+    if (files.empty()) {
+        log_info() << "No configuration files found: " << m_source;
+        return;
+    }
+
+    for (std::string file : files) {
+        std::string filePath = buildPath(m_source, file);
+
+        if (!isJsonFile(filePath)) {
+            errorMessage = "The file is not a json file: " + filePath;
+            log_debug() << errorMessage;
+            throw ServiceManifestParseError(errorMessage);
+        }
+
+        loadServiceManifestFile(filePath);
+    }
+}
+
+void ServiceManifestFileLoader::loadServiceManifestFile(const std::string &filePath)
+{
+    std::string errorMessage;
+
+    json_error_t error;
+    json_t *fileroot = json_load_file(m_source.c_str(), 0, &error);
+
+    if (nullptr == fileroot) {
+        errorMessage = "Could not parse the Service Manifest: "
+            + filePath + ":" + std::to_string(error.line) + " : " + std::string(error.text);
+        log_error() << errorMessage;
+        throw ServiceManifestParseError(errorMessage);
+    } else if (!json_is_object(fileroot)) {
+        errorMessage = "The Service Manifest does not contain a json object: "
+            + filePath + ":" + std::to_string(error.line) + " : " + std::string(error.text);
+        log_error() << errorMessage;
+        throw ServiceManifestParseError(errorMessage);
+    }
+
+    log_debug() << "Loaded Service Manifest: " << filePath;
+
+    m_content.push_back(fileroot);
+}
+
+
+/*
+ * Help functions
+ */
+
+std::vector<std::string> ServiceManifestFileLoader::fileList()
+{
+    std::vector<std::string> files;
+
+    DIR *dirFile = opendir(m_source.c_str());
```

Comment:
Consider using the Dir class from Glibmm instead: https://developer.gnome.org/glibmm/stable/classGlib_1_1Dir.html

Should make traversing easier etc and provides a higher abstraction level than the C system API.